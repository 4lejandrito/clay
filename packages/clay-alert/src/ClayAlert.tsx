/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as React from 'react';
import classNames from 'classnames';
import Icon from '@clayui/icon';

type DisplayType = 'danger' | 'info' | 'success' | 'warning';

interface ClayAlertProps extends React.HTMLAttributes<HTMLDivElement> {
	/**
	 * Callback function for when the 'x' is clicked.
	 */
	onClose?: () => void;

	/**
	 * Determines the style of the alert.
	 */
	displayType?: DisplayType;

	/**
	 * Flag to indicate the stripe variant.
	 */
	stripe?: boolean;

	/**
	 * The summary of the Alert, often is something like 'Error' or 'Info'.
	 */
	title: string;

	/**
	 * Path to the spritemap that Icon should use when referencing symbols.
	 */
	spritemap?: string;
}

interface ToastContainerProps extends React.HTMLAttributes<HTMLDivElement> {
	/**
	 * Chilren of the ToastContainer must be a ClayAlert
	 */
	children?:
		| React.ReactElement<ClayAlertProps>
		| React.ReactElement<ClayAlertProps>[];
}

const ICON_MAP = {
	danger: 'exclamation-full',
	info: 'info-circle',
	success: 'check-circle-full',
	warning: 'warning-full',
};

const ClayAlert: React.FunctionComponent<ClayAlertProps> & {
	ToastContainer: React.FunctionComponent<ToastContainerProps>;
} = ({
	children,
	className,
	displayType = 'info',
	onClose,
	spritemap,
	stripe,
	title,
	...otherProps
}) => {
	const ConditionalContainer: React.FunctionComponent<{}> = ({children}) =>
		stripe ? <div className="container">{children}</div> : <>{children}</>;

	return (
		<div
			{...otherProps}
			className={classNames(className, 'alert', {
				'alert-dismissible': onClose,
				'alert-fluid': stripe,
				[`alert-${displayType}`]: displayType,
			})}
			role="alert"
		>
			<ConditionalContainer>
				<span className="alert-indicator">
					<Icon
						spritemap={spritemap}
						symbol={ICON_MAP[displayType]}
					/>
				</span>

				<strong className="lead">{title}</strong>

				{children}

				{onClose && (
					<button
						aria-label="Close"
						className="close"
						onClick={onClose}
						type="button"
					>
						<Icon spritemap={spritemap} symbol="times" />
					</button>
				)}
			</ConditionalContainer>
		</div>
	);
};

ClayAlert.ToastContainer = ({children, className, ...otherProps}) => {
	return (
		<div
			{...otherProps}
			className={classNames(className, 'alert-container container')}
		>
			<div className="alert-notifications alert-notifications-fixed">
				{children}
			</div>
		</div>
	);
};

export default ClayAlert;
