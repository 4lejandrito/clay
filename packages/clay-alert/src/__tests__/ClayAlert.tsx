/**
 * Â© 2019 Liferay, Inc. <https://liferay.com>
 *
 * SPDX-License-Identifier: BSD-3-Clause
 */

import * as React from 'react';
import * as TestRenderer from 'react-test-renderer';
import ClayAlert from '../ClayAlert';

describe('ClayAlert', () => {
	it('renders', () => {
		const testRenderer = TestRenderer.create(<ClayAlert title="Hello!" />);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('renders as a different type', () => {
		const testRenderer = TestRenderer.create(
			<ClayAlert displayType="danger" title="Hello!" />
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('renders as `stripe` variant', () => {
		const testRenderer = TestRenderer.create(
			<ClayAlert stripe title="Hello!" />
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('renders with an icon for closing', () => {
		const testRenderer = TestRenderer.create(
			<ClayAlert onClose={() => {}} title="Hello!" />
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('renders with a title and a message with markup', () => {
		const testRenderer = TestRenderer.create(
			<ClayAlert title="Hello!">
				<span>{'test'}</span>
			</ClayAlert>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});

	it('renders with ToastContainer as a wrapper ', () => {
		const testRenderer = TestRenderer.create(
			<ClayAlert.ToastContainer>
				<ClayAlert title="One!" />
				<ClayAlert title="Two!" />
				<ClayAlert title="Three!" />
			</ClayAlert.ToastContainer>
		);

		expect(testRenderer.toJSON()).toMatchSnapshot();
	});
});
