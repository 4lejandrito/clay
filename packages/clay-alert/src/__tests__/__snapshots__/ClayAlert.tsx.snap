// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClayAlert renders 1`] = `
<div
  className="alert alert-info"
  role="alert"
>
  <span
    className="alert-indicator"
  >
    <svg
      className="lexicon-icon lexicon-icon-info-circle"
      role="presentation"
    >
      <use
        xlinkHref="#info-circle"
      />
    </svg>
  </span>
  <strong
    className="lead"
  >
    Hello!
  </strong>
</div>
`;

exports[`ClayAlert renders as \`stripe\` variant 1`] = `
<div
  className="alert alert-fluid alert-info"
  role="alert"
>
  <div
    className="container"
  >
    <span
      className="alert-indicator"
    >
      <svg
        className="lexicon-icon lexicon-icon-info-circle"
        role="presentation"
      >
        <use
          xlinkHref="#info-circle"
        />
      </svg>
    </span>
    <strong
      className="lead"
    >
      Hello!
    </strong>
  </div>
</div>
`;

exports[`ClayAlert renders as a different type 1`] = `
<div
  className="alert alert-danger"
  role="alert"
>
  <span
    className="alert-indicator"
  >
    <svg
      className="lexicon-icon lexicon-icon-exclamation-full"
      role="presentation"
    >
      <use
        xlinkHref="#exclamation-full"
      />
    </svg>
  </span>
  <strong
    className="lead"
  >
    Hello!
  </strong>
</div>
`;

exports[`ClayAlert renders with ToastContainer as a wrapper  1`] = `
<div
  className="alert-container container"
>
  <div
    className="alert-notifications alert-notifications-fixed"
  >
    <div
      className="alert alert-info"
      role="alert"
    >
      <span
        className="alert-indicator"
      >
        <svg
          className="lexicon-icon lexicon-icon-info-circle"
          role="presentation"
        >
          <use
            xlinkHref="#info-circle"
          />
        </svg>
      </span>
      <strong
        className="lead"
      >
        One!
      </strong>
    </div>
    <div
      className="alert alert-info"
      role="alert"
    >
      <span
        className="alert-indicator"
      >
        <svg
          className="lexicon-icon lexicon-icon-info-circle"
          role="presentation"
        >
          <use
            xlinkHref="#info-circle"
          />
        </svg>
      </span>
      <strong
        className="lead"
      >
        Two!
      </strong>
    </div>
    <div
      className="alert alert-info"
      role="alert"
    >
      <span
        className="alert-indicator"
      >
        <svg
          className="lexicon-icon lexicon-icon-info-circle"
          role="presentation"
        >
          <use
            xlinkHref="#info-circle"
          />
        </svg>
      </span>
      <strong
        className="lead"
      >
        Three!
      </strong>
    </div>
  </div>
</div>
`;

exports[`ClayAlert renders with a title and a message with markup 1`] = `
<div
  className="alert alert-info"
  role="alert"
>
  <span
    className="alert-indicator"
  >
    <svg
      className="lexicon-icon lexicon-icon-info-circle"
      role="presentation"
    >
      <use
        xlinkHref="#info-circle"
      />
    </svg>
  </span>
  <strong
    className="lead"
  >
    Hello!
  </strong>
  <span>
    test
  </span>
</div>
`;

exports[`ClayAlert renders with an icon for closing 1`] = `
<div
  className="alert alert-dismissible alert-info"
  role="alert"
>
  <span
    className="alert-indicator"
  >
    <svg
      className="lexicon-icon lexicon-icon-info-circle"
      role="presentation"
    >
      <use
        xlinkHref="#info-circle"
      />
    </svg>
  </span>
  <strong
    className="lead"
  >
    Hello!
  </strong>
  <button
    aria-label="Close"
    className="close"
    onClick={[Function]}
    type="button"
  >
    <svg
      className="lexicon-icon lexicon-icon-times"
      role="presentation"
    >
      <use
        xlinkHref="#times"
      />
    </svg>
  </button>
</div>
`;
